source/slang/parser.cpp

```diff
@@ -598,10 +646,57 @@ namespace Slang
         AddModifier(&modifierLink, modifier);
     }
 
+    // 
+    // '::'? identifier ('::' identifier)* 
+    static Token parseAttributeName(Parser* parser)
+    {
+        const SourceLoc scopedIdSourceLoc = parser->tokenReader.PeekLoc();
+
+        // Strip initial :: if there is one
+        const TokenType initialTokenType = parser->tokenReader.PeekTokenType();
+        if (initialTokenType == TokenType::Scope)
+        {
+            parser->ReadToken(TokenType::Scope); 
+        }
+
+        const Token firstIdentifier = parser->ReadToken(TokenType::Identifier);
+        if (initialTokenType != TokenType::Scope && parser->tokenReader.PeekTokenType() != TokenType::Scope)
+        {
+            return firstIdentifier;
+        }
+
+        // Build up scoped string
+        StringBuilder scopedIdentifierBuilder;
+        if (initialTokenType == TokenType::Scope)
+        {
+            scopedIdentifierBuilder.Append('_'); 
+        }
+        scopedIdentifierBuilder.Append(firstIdentifier.Content);
+
+        while (parser->tokenReader.PeekTokenType() == TokenType::Scope)
+        {
+            parser->ReadToken(TokenType::Scope);
+            scopedIdentifierBuilder.Append('_'); 
+            
+            const Token nextIdentifier(parser->ReadToken(TokenType::Identifier));
+            scopedIdentifierBuilder.Append(nextIdentifier.Content);
+        }
+
+        // Make a 'token'
+        const String scopedIdentifier(scopedIdentifierBuilder.ToString());
+        Token token(TokenType::Identifier, scopedIdentifier, scopedIdSourceLoc);
+        token.ptrValue = getName(parser, token.Content);
```

Comment:
I'm kind of antsy about constructing a `Token` on the fly here that doesn't accurately reflect the source code as the user wrote it (or after preprocessing). It is a small example of the compiler lying (to itself), but in this case it is the best short-term answer.

A better long-term answer would have us introduce something like a `QualifiedName` into the AST and use that for attribute names instead of a single `Token`.

Reply:
Agree that it's not great for the reasons you discuss. And yes something that better understood scoping and the structure would be better.